cluster.c

```diff
@@ -623,13 +741,22 @@ RRStatus ShardingInfoAddShardGroup(RedisRaftCtx *rr, ShardGroup *new_sg)
     if (ShardingInfoValidateShardGroup(rr, new_sg) != RR_OK)
         return RR_ERROR;
 
-    si->shard_groups_num++;
+    unsigned int * idx = RedisModule_Alloc(sizeof(unsigned int));
+    *idx = si->shard_groups_num;
+    if (RedisModule_DictSetC(si->shard_group_map, new_sg->id, sizeof(new_sg->id), idx) != REDISMODULE_OK) {
+        RedisModule_Free(idx);
+        return RR_ERROR;
+    }
+
+    unsigned int sg_idx = si->shard_groups_num++;
     si->shard_groups = RedisModule_Realloc(si->shard_groups, sizeof(ShardGroup *) * si->shard_groups_num);
 
-    ShardGroup *sg = si->shard_groups[si->shard_groups_num-1] = RedisModule_Alloc(sizeof(ShardGroup));
-    sg->id = si->shard_groups_num;
-    sg->start_slot = new_sg->start_slot;
-    sg->end_slot = new_sg->end_slot;
+    ShardGroup *sg = si->shard_groups[sg_idx] = RedisModule_Alloc(sizeof(ShardGroup));
+    strncpy(sg->id, new_sg->id, sizeof(sg->id));
```

Comment:
Same as above.

Reply:
note quite sure I understand

Reply:
how would we avoid the strncpy in this case?

Reply:
It's a fixed length buffer, no need to treat it as a null terminated string. And if it was null terminated, but longer than N, `strncpy` is a risky function that is best avoided because it will produce non-null-terminated strings in the destination.

Reply:
in our case, it should never be non null terminated, as are are always strncpy(LEN) when the fixed sized buffer is LEN+1

In practice, I think it can be considered fix size, but I'm wondering if we should protect against users passing in wrong sized strings in.  perhaps better in those cases to verify up front?

Reply:
Remember `strncpy` treats `n` as the size of the buffer and will not append a null byte if `source length == n`.

Reply:
the buffer is 64+1, the str should be capped at 64.  should we just throw an assert in somewhere verifying that?

Reply:
would this be sufficient?

```
    if (strlen(new_sg) >= sizeof(sg->id)) {
        return RR_ERROR;
    }
    strncpy(sg->id, new_sg->id, sizeof(sg->id));
```